docs: DOC-615: deephaven design v2 (#7715)
deephaven/deephaven-core: docs/groovy/conceptual/deephaven-design.md

A table may share any of its `ColumnSources` with any other table in its update graph that uses the same row key space or whose row key space can be _redirected_ onto the same row key space. A single parent table is responsible for updating the contents of a given `ColumnSource` for itself and any descendants that inherited that `ColumnSource`. Operations that allow this sharing without redirection include [`where`](../reference/table-operations/filter/where.md), as well as any column transformation that passes through columns unchanged or simply renamed, and some join operations with respect to the left-hand table’s `ColumnSources`. Operations that allow `ColumnSource` sharing with redirection include [`sort`](../reference/table-operations/sort/sort.md), as well as most joins with respect to the right-hand table’s `ColumnSources`.

By itself, this sharing capability represents an important optimization that avoids some data processing or copying work. When considered in an _updating_ query engine, it should be clear that this avoidance extends to each update cycle, paying dividends for the lifetime of a query.

> **Performance impact**: Shared `ColumnSources` can reduce memory footprint by 50-90% for queries with multiple filtered or joined views of the same source data. For example, five different `where` filters on a 10-column table share all 10 `ColumnSources`, storing only five different `RowSets` instead of duplicating 50 columns.

Reviewer: ```suggestion
**Performance impact**: Shared `ColumnSource`s can reduce memory footprint by 50-90% for queries with multiple filtered or joined views of the same source data. For example, five different `where` filters on a 10-column table share all 10 `ColumnSource`s, storing only five different `RowSet`s instead of duplicating 50 columns.
```

A table may share any of its ColumnSources with any other table in its update graph that uses the same row key space or whose row key space can be _redirected_ onto the same row key space. A single parent table is responsible for updating the contents of a given ColumnSource for itself and any descendants that inherited that ColumnSource. Operations that allow this sharing without redirection include [`where`](../reference/table-operations/filter/where.md), as well as any column transformation that passes through columns unchanged or simply renamed, and some join operations with respect to the left-hand table’s ColumnSources. Operations that allow ColumnSource sharing with redirection include [`sort`](../reference/table-operations/sort/sort.md), as well as most joins with respect to the right-hand table’s ColumnSources.

By itself, this sharing capability represents an important optimization that avoids some data processing or copying work. When considered in an _updating_ query engine, it should be clear that this avoidance extends to each update cycle, paying dividends for the lifetime of a query.

> **Performance impact**: Shared ColumnSources can reduce memory footprint by 50-90% for queries with multiple filtered or joined views of the same source data. For example, five different `where()` filters on a 10-column table share all 10 ColumnSources, storing only five different RowSets instead of duplicating 50 columns.

Reviewer: ```suggestion
**Performance impact**: Shared ColumnSources can reduce memory footprint by 50-90% for queries with multiple filtered or joined views of the same source data. For example, five different `where` filters on a 10-column table share all 10 ColumnSources, storing only five different RowSets instead of duplicating 50 columns.
```

link on `where`?

Chunks are currently implemented as regions of native Java arrays, with implementations for each primitive type as well as an additional one for Object references. In the future we might switch to use [Apache Arrow](https://arrow.apache.org/docs/index.html) [ValueVectors](https://arrow.apache.org/java/current/vector.html) (called Arrays in the C++ implementation) or another high-performance implementation.

### Columnar hash tables

Building further on our chunk-oriented engine design, Deephaven [join](../how-to-guides/joins-exact-relational.md) and [aggregation](../how-to-guides/combined-aggregations.md) operations frequently rely on a columnar hash table implementation that allows chunks of single or multi-element keys to be hashed and inserted or probed in bulk operations where data patterns permit. Each component of this operational building block attempts to enable vectorization and avoid virtual method lookups by relying on chunk-oriented kernels, whether for hashing data, sorting data, looking for runs of identical values, and so on.

### `RowSet` implementations

> Another area of the Deephaven query engine that merits specific attention is our [`RowSet`](#tables-designed-for-sharing-and-updating) implementation. In practice, `RowSets` switch between a number of implementation options depending on size and sparsity. Currently, there are three in use:

Reviewer: ```suggestion
Another area of the Deephaven query engine that merits specific attention is our [`RowSet`](#tables-designed-for-sharing-and-updating) implementation. In practice, `RowSet`s switch between a number of implementation options depending on size and sparsity. Currently, there are three in use:
```

Consider operations such as union, set difference, or intersect. The system can work on the corresponding containers without caring about boundaries between them. For instance, the position for `RowSet` key 65,537 can _only_ be in the container where n = 1, so Deephaven always knows exactly where to look.

For large datasets, RSP is usually much faster at these operations. If the system can work in pieces at a time for each operation and guarantee that the pieces on each `RowSet` align in terms of values, they can be worked on more efficiently without having to worry about iterators crossing individual array boundaries at different points for each `RowSet`.

RSP also compactly represents values. It would be inefficient to write out every single value in a big range - let's say all of the values in the set `[ 0, 1000 * 2^16 ]`. Deephaven would have to create 10,000 containers, with all keys in each of them present. Instead, Deephaven has a compact way to represent continuous ranges of values that can span full blocks in `[ n * 2^16, m * 2^16 - 1 ]` space, for some value of `n` and `m` where `m > n`.

> Nevertheless, small `RowSets` are more expensive in RSP because they are still comprised of an array of containers first and then the containers themselves. This matters for queries that need to create one `RowSet` per row (e.g., joins). This issue is alleviated by creating a special case for single-range `RowSets`.

Reviewer: Move backtick
